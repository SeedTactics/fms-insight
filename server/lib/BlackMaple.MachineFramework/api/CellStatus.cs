/* Copyright (c) 2023, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#nullable enable

using System;
using System.Runtime.Serialization;
using System.Collections.Immutable;
using Germinate;

namespace BlackMaple.MachineFramework
{
  [DataContract]
  public record QueueSize
  {
    //once an output queue grows to this size, stop unloading parts
    //and keep them in the buffer inside the cell
    [DataMember(IsRequired = false, EmitDefaultValue = false)]
    public int? MaxSizeBeforeStopUnloading { get; init; }
  }

  [DataContract, Draftable]
  public record CurrentStatus
  {
    [DataMember(IsRequired = true)]
    public required DateTime TimeOfCurrentStatusUTC { get; init; }

    [DataMember(Name = "Jobs", IsRequired = true)]
    public required ImmutableDictionary<string, MachineFramework.ActiveJob> Jobs { get; init; }

    [DataMember(Name = "Pallets", IsRequired = true)]
    public required ImmutableDictionary<string, PalletStatus> Pallets { get; init; }

    [DataMember(Name = "Material", IsRequired = true)]
    public required ImmutableList<InProcessMaterial> Material { get; init; }

    [DataMember(Name = "Alarms", IsRequired = true)]
    public required ImmutableList<string> Alarms { get; init; }

    [DataMember(Name = "Queues", IsRequired = true)]
    public required ImmutableDictionary<string, QueueSize> QueueSizes { get; init; }

    // The following is only filled in if the machines move to the load station
    // instead of the pallets moving to the machine
    [DataMember(Name = "MachineLocations", IsRequired = false, EmitDefaultValue = false)]
    public ImmutableList<MachineLocation>? MachineLocations { get; init; }

    public static CurrentStatus operator %(CurrentStatus s, Action<ICurrentStatusDraft> f) => s.Produce(f);
  }

  [DataContract]
  public record JobAndDecrementQuantity
  {
    [DataMember(IsRequired = true)]
    public required long DecrementId { get; init; }

    [DataMember(IsRequired = true)]
    public required string JobUnique { get; init; }

    [DataMember(IsRequired = true)]
    public required DateTime TimeUTC { get; init; }

    [DataMember(IsRequired = true)]
    public required string Part { get; init; }

    [DataMember(IsRequired = true)]
    public required int Quantity { get; init; }
  }

  [DataContract]
  public record HistoricData
  {
    [DataMember(IsRequired = true)]
    public required ImmutableDictionary<string, MachineFramework.HistoricJob> Jobs { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableList<SimulatedStationUtilization> StationUse { get; init; }
  }

  [DataContract]
  public record PlannedSchedule
  {
    [DataMember(IsRequired = true)]
    public required string LatestScheduleId { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableList<MachineFramework.HistoricJob> Jobs { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableDictionary<string, int> ExtraParts { get; init; }

    [DataMember(IsRequired = false)]
    public ImmutableList<Workorder>? CurrentUnfilledWorkorders { get; init; }
  }
}
