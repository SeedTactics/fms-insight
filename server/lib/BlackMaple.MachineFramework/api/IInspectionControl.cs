/* Copyright (c) 2017, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;

namespace BlackMaple.MachineWatchInterface
{
  [Serializable, DataContract]
  public struct InspectCount
  {
    [DataMember(IsRequired = true)] public string Counter;
    [DataMember(IsRequired = true)] public int Value;
    [DataMember(IsRequired = true)] public DateTime LastUTC;
  }

  public interface IInspectionControl : IDisposable
  {
    ///Forces the given materialID to be inspected
    void ForceInspection(long materialID, string inspectionType);

    ///Forces the next piece that runs on the given location to be inspected
    void NextPieceInspection(MachineWatchInterface.PalletLocation palLoc, string inspType);

    //allow editing of counts.  Uses lists of InspectCount structure
    List<InspectCount> LoadInspectCounts();
    void SetInspectCounts(IEnumerable<InspectCount> countUpdates);
  }
}
