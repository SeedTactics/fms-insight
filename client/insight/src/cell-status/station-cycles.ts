/* Copyright (c) 2021, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { atom, RecoilValueReadOnly, TransactionInterface_UNSTABLE } from "recoil";
import { IEditMaterialInLogEvents, ILogEntry, ILogMaterial, LogType } from "../network/api";
import { conduit } from "../util/recoil-util";
import type { ServerEventAndTime } from "./loading";
import { LazySeq } from "../util/lazyseq";
import {
  activeMinutes,
  EstimatedCycleTimes,
  last30EstimatedCycleTimes,
  LogEntryWithSplitElapsed,
  PartAndStationOperation,
  specificMonthEstimatedCycleTimes,
  splitElapsedLoadTime,
} from "./estimated-cycle-times";
import { durationToMinutes } from "../util/parseISODuration";
import { addDays } from "date-fns";
import { emptyIMap, IMap } from "../util/imap";

export interface PartCycleData {
  readonly x: Date;
  readonly y: number; // cycle time in minutes
  readonly cntr: number;
  readonly part: string;
  readonly process: number;
  readonly stationGroup: string;
  readonly stationNumber: number;
  readonly operation: string;
  readonly pallet: string;
  readonly activeMinutes: number; // active time in minutes
  readonly medianCycleMinutes: number;
  readonly MAD_aboveMinutes: number;
  readonly isLabor: boolean;
  readonly material: ReadonlyArray<Readonly<ILogMaterial>>;
  readonly operator: string;
}

export type StationCyclesByCntr = IMap<number, PartCycleData>;

export function stat_name_and_num(stationGroup: string, stationNumber: number): string {
  if (stationGroup.startsWith("Inspect")) {
    return stationGroup;
  } else {
    return stationGroup + " #" + stationNumber.toString();
  }
}

export function splitElapsedLoadTimeAmongCycles(
  cycles: LazySeq<PartCycleData>
): LazySeq<LogEntryWithSplitElapsed<PartCycleData>> {
  return splitElapsedLoadTime(
    cycles,
    (c) => c.stationNumber,
    (c) => c.x,
    (c) => c.y,
    (c) => c.activeMinutes
  );
}

const last30StationCyclesRW = atom<StationCyclesByCntr>({
  key: "last30StationCycles",
  default: emptyIMap(),
});
export const last30StationCycles: RecoilValueReadOnly<StationCyclesByCntr> = last30StationCyclesRW;

const specificMonthStationCyclesRW = atom<StationCyclesByCntr>({
  key: "specificMonthStationCycles",
  default: emptyIMap(),
});
export const specificMonthStationCycles: RecoilValueReadOnly<StationCyclesByCntr> = specificMonthStationCyclesRW;

function convertLogToCycle(estimatedCycleTimes: EstimatedCycleTimes, cycle: ILogEntry): [number, PartCycleData] | null {
  if (
    cycle.startofcycle ||
    (cycle.type !== LogType.LoadUnloadCycle && cycle.type !== LogType.MachineCycle) ||
    cycle.loc === ""
  ) {
    return null;
  }
  const part = cycle.material.length > 0 ? cycle.material[0].part : "";
  const proc = cycle.material.length > 0 ? cycle.material[0].proc : 1;
  const stats =
    cycle.material.length > 0 ? estimatedCycleTimes.get(PartAndStationOperation.ofLogCycle(cycle)) : undefined;
  const elapsed = durationToMinutes(cycle.elapsed);
  return [
    cycle.counter,
    {
      x: cycle.endUTC,
      y: elapsed,
      cntr: cycle.counter,
      activeMinutes: activeMinutes(cycle, stats),
      medianCycleMinutes: (stats?.medianMinutesForSingleMat ?? 0) * cycle.material.length,
      MAD_aboveMinutes: stats?.MAD_aboveMinutes ?? 0,
      part: part,
      process: proc,
      pallet: cycle.pal,
      material: cycle.material,
      isLabor: cycle.type === LogType.LoadUnloadCycle,
      stationGroup: cycle.loc,
      stationNumber: cycle.locnum,
      operation: cycle.type === LogType.LoadUnloadCycle ? cycle.result : cycle.program,
      operator: cycle.details ? cycle.details.operator || "" : "",
    },
  ];
}

function process_swap(swap: Readonly<IEditMaterialInLogEvents>, partCycles: StationCyclesByCntr): StationCyclesByCntr {
  for (const changed of swap.editedEvents) {
    const c = partCycles.get(changed.counter);
    if (c !== undefined) {
      const newC = { ...c, material: changed.material };
      partCycles = partCycles.set(changed.counter, newC);
    }
  }
  return partCycles;
}

export const setLast30StationCycles = conduit<ReadonlyArray<Readonly<ILogEntry>>>(
  (t: TransactionInterface_UNSTABLE, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    const estimatedCycleTimes = t.get(last30EstimatedCycleTimes);
    t.set(last30StationCyclesRW, (oldCycles) =>
      oldCycles.size === 0
        ? LazySeq.ofIterable(log)
            .collect((c) => convertLogToCycle(estimatedCycleTimes, c))
            .toIMap((x) => x)
        : oldCycles.append(
            LazySeq.ofIterable(log).collect((c) => convertLogToCycle(estimatedCycleTimes, c)),
            (_, b) => b
          )
    );
  }
);

export const updateLast30StationCycles = conduit<ServerEventAndTime>(
  (t: TransactionInterface_UNSTABLE, { evt, now, expire }: ServerEventAndTime) => {
    if (evt.logEntry && evt.logEntry.type === LogType.InvalidateCycle) {
      const cntrs = evt.logEntry.details?.["EditedCounters"];
      const invalidatedCycles = cntrs ? new Set(cntrs.split(",").map((i) => parseInt(i))) : new Set<number>();

      if (invalidatedCycles.size > 0) {
        t.set(last30StationCyclesRW, (cycles) => {
          for (const invalid of invalidatedCycles) {
            const c = cycles.get(invalid);
            if (c !== undefined) {
              cycles = cycles.set(invalid, { ...c, activeMinutes: 0 });
            }
          }
          return cycles;
        });
      }
    } else if (evt.logEntry) {
      const estimatedCycleTimes = t.get(last30EstimatedCycleTimes);
      const converted = convertLogToCycle(estimatedCycleTimes, evt.logEntry);
      if (!converted) return;

      t.set(last30StationCyclesRW, (cycles) => {
        if (expire) {
          const thirtyDaysAgo = addDays(now, -30);
          cycles = cycles.bulkDelete((_, e) => e.x < thirtyDaysAgo);
        }

        cycles = cycles.set(converted[0], converted[1]);

        return cycles;
      });
    } else if (evt.editMaterialInLog) {
      const edit = evt.editMaterialInLog;
      t.set(last30StationCyclesRW, (oldCycles) => process_swap(edit, oldCycles));
    }
  }
);

export const setSpecificMonthStationCycles = conduit<ReadonlyArray<Readonly<ILogEntry>>>(
  (t: TransactionInterface_UNSTABLE, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    const estimatedCycleTimes = t.get(specificMonthEstimatedCycleTimes);
    t.set(
      specificMonthStationCyclesRW,
      LazySeq.ofIterable(log)
        .collect((c) => convertLogToCycle(estimatedCycleTimes, c))
        .toIMap((x) => x)
    );
  }
);
