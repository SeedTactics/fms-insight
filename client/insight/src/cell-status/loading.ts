/* Copyright (c) 2021, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { atom, RecoilValueReadOnly, TransactionInterface_UNSTABLE } from "recoil";
import { ICurrentStatus, IHistoricData, ILogEntry, IServerEvent } from "../network/api.js";
import { conduit } from "../util/recoil-util.js";
import { LazySeq } from "@seedtactics/immutable-collections";

import * as simProd from "./sim-production.js";
import * as simUse from "./sim-station-use.js";
import * as schJobs from "./scheduled-jobs.js";
import * as buffers from "./buffers.js";
import * as currentSt from "./current-status.js";
import * as insp from "./inspections.js";
import * as mats from "./material-summary.js";
import * as names from "./names.js";
import * as estimated from "./estimated-cycle-times.js";
import * as tool from "./tool-usage.js";
import * as palCycles from "./pallet-cycles.js";
import * as statCycles from "./station-cycles.js";
import * as toolReplace from "./tool-replacements.js";

export interface ServerEventAndTime {
  readonly evt: Readonly<IServerEvent>;
  readonly now: Date;
  readonly expire: boolean;
}

const lastEventCounterRW = atom<number | null>({
  key: "log-last-event-counter",
  default: null,
});
export const lastEventCounter: RecoilValueReadOnly<number | null> = lastEventCounterRW;

export const onServerEvent = conduit<ServerEventAndTime>(
  (t: TransactionInterface_UNSTABLE, evt: ServerEventAndTime) => {
    simProd.updateLast30JobProduction.transform(t, evt);
    simUse.updateLast30SimStatUse.transform(t, evt);
    schJobs.updateLast30Jobs.transform(t, evt);
    buffers.updateLast30Buffer.transform(t, evt);
    currentSt.updateCurrentStatus.transform(t, evt);
    mats.updateLast30MatSummary.transform(t, evt);
    insp.updateLast30Inspections.transform(t, evt);
    names.updateNames.transform(t, evt);
    tool.updateLast30ToolUse.transform(t, evt);
    palCycles.updateLast30PalletCycles.transform(t, evt);
    toolReplace.updateLastToolReplacements.transform(t, evt);
    statCycles.updateLast30StationCycles.transform(t, evt);

    if (evt.evt.logEntry) {
      const newCntr = evt.evt.logEntry.counter;
      t.set(lastEventCounterRW, (old) => (old === null ? old : Math.max(old, newCntr)));
    }
  }
);

export const onLoadLast30Jobs = conduit<Readonly<IHistoricData>>(
  (t: TransactionInterface_UNSTABLE, historicData: Readonly<IHistoricData>) => {
    simUse.setLast30SimStatUse.transform(t, historicData);
    simProd.setLast30JobProduction.transform(t, historicData);
    schJobs.setLast30Jobs.transform(t, historicData);
    names.setNamesFromLast30Jobs.transform(t, historicData);
  }
);

export const onLoadLast30Log = conduit<ReadonlyArray<Readonly<ILogEntry>>>(
  (t: TransactionInterface_UNSTABLE, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    estimated.setLast30EstimatedCycleTimes.transform(t, log);
    buffers.setLast30Buffer.transform(t, log);
    insp.setLast30Inspections.transform(t, log);
    mats.setLast30MatSummary.transform(t, log);
    names.setNamesFromLast30Evts.transform(t, log);
    tool.setLast30ToolUse.transform(t, log);
    palCycles.setLast30PalletCycles.transform(t, log);
    toolReplace.setLast30ToolReplacements.transform(t, log);
    statCycles.setLast30StationCycles.transform(t, log);

    const newCntr = LazySeq.of(log).maxBy((x) => x.counter)?.counter ?? null;
    t.set(lastEventCounterRW, (oldCntr) => (oldCntr === null ? newCntr : Math.max(oldCntr, newCntr ?? -1)));
  }
);

export const onLoadCurrentSt = conduit<Readonly<ICurrentStatus>>(
  (t: TransactionInterface_UNSTABLE, curSt: Readonly<ICurrentStatus>) => {
    currentSt.setCurrentStatus.transform(t, curSt);
  }
);

export const onLoadSpecificMonthJobs = conduit<Readonly<IHistoricData>>(
  (t: TransactionInterface_UNSTABLE, historicData: Readonly<IHistoricData>) => {
    simUse.setSpecificMonthSimStatUse.transform(t, historicData);
    simProd.setSpecificMonthJobProduction.transform(t, historicData);
    schJobs.updateSpecificMonthJobs.transform(t, historicData);
  }
);

export const onLoadSpecificMonthLog = conduit<ReadonlyArray<Readonly<ILogEntry>>>(
  (t: TransactionInterface_UNSTABLE, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    estimated.setSpecificMonthEstimatedCycleTimes.transform(t, log);
    buffers.setSpecificMonthBuffer.transform(t, log);
    insp.setSpecificMonthInspections.transform(t, log);
    mats.setSpecificMonthMatSummary.transform(t, log);
    palCycles.setSpecificMonthPalletCycles.transform(t, log);
    toolReplace.setSpecificMonthToolReplacements.transform(t, log);
    statCycles.setSpecificMonthStationCycles.transform(t, log);
  }
);
