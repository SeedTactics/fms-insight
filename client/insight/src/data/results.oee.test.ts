/* Copyright (c) 2020, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { addHours, differenceInMinutes, addMinutes } from "date-fns";

import { fakeCycle, fakeMaterial } from "../../test/events.fake.js";
import { ILogEntry, LogType } from "../network/api.js";
import { LazySeq } from "@seedtactics/immutable-collections";
import {
  binActiveCyclesByDayAndStat,
  binOccupiedCyclesByDayAndStat,
  buildOeeHeatmapTable,
} from "./results.oee.js";
import { onLoadLast30Log } from "../cell-status/loading.js";
import { last30StationCycles } from "../cell-status/station-cycles.js";
import { it, expect } from "vitest";
import { toRawJs } from "../../test/to-raw-js.js";
import { createStore } from "jotai";

it("bins actual cycles by day", () => {
  const now = new Date(2018, 2, 5); // midnight in local time
  const nowChicago = new Date(Date.UTC(2018, 2, 5, 6, 0, 0)); // America/Chicago time
  const minOffset = differenceInMinutes(nowChicago, now);

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle({ time: now, machineTime: 30, counter: 100 }),
    fakeCycle({ time: addHours(now, -3), machineTime: 20, counter: 200 }),
    fakeCycle({ time: addHours(now, -15), machineTime: 15, counter: 300 }),
    LazySeq.ofRange(1, 3)
      .map((i) => {
        const material = [fakeMaterial()];
        return {
          counter: 1,
          material,
          pal: 2,
          type: LogType.LoadUnloadCycle,
          startofcycle: false,
          endUTC: addHours(now, -4), // same time for both cycles
          loc: "L/U",
          locnum: 3,
          result: i === 1 ? "LOAD" : "UNLOAD",
          program: i === 1 ? "LOAD" : "UNLOAD",
          elapsed: "PT8M",
          active: "PT3M",
        };
      })
      .toRArray(),
  );

  const snapshot = createStore();
  snapshot.set(onLoadLast30Log, evts);
  const cycles = snapshot.get(last30StationCycles);

  let byDayAndStat = binActiveCyclesByDayAndStat(cycles.valuesToLazySeq());

  // update day to be in Chicago timezone
  // This is because the snapshot formats the day as a UTC time in Chicago timezone
  // Note this is after cycles are binned, which is correct since cycles are generated using
  // now in local time and then binned in local time.  Just need to update the date before
  // comparing with the snapshot
  byDayAndStat = byDayAndStat
    .toLazySeq()
    .toHashMap(([dayAndStat, val]) => [dayAndStat.adjustDay((d) => addMinutes(d, minOffset)), val]);

  expect(toRawJs(byDayAndStat)).toMatchSnapshot("cycles binned by day and station");

  byDayAndStat = binOccupiedCyclesByDayAndStat(cycles.valuesToLazySeq());
  byDayAndStat = byDayAndStat
    .toLazySeq()
    .toHashMap(([dayAndStat, val]) => [dayAndStat.adjustDay((d) => addMinutes(d, minOffset)), val]);

  expect(toRawJs(byDayAndStat)).toMatchSnapshot("occupied cycles binned by day and station");
});

it("creates points clipboard table", () => {
  // table formats columns in local time, so no need to convert to a specific timezone
  const now = new Date(2018, 2, 5); // midnight in local time

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle({ time: now, machineTime: 30, counter: 100 }),
    fakeCycle({ time: addHours(now, -3), machineTime: 20, counter: 200 }),
    fakeCycle({ time: addHours(now, -15), machineTime: 15, counter: 300 }),
  );

  const snapshot = createStore();
  snapshot.set(onLoadLast30Log, evts);
  const cycles = snapshot.get(last30StationCycles);

  const byDayAndStat = binActiveCyclesByDayAndStat(cycles.valuesToLazySeq());

  const points = LazySeq.of(byDayAndStat)
    .map(([dayAndStat, val]) => ({
      x: dayAndStat.day,
      y: dayAndStat.station,
      label: val.toString(),
    }))
    .toRArray();

  const table = document.createElement("div");
  table.innerHTML = buildOeeHeatmapTable("Station", points);
  expect(table).toMatchSnapshot("clipboard table");
});
