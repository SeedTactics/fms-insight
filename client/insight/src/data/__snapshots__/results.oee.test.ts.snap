// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bins actual cycles by day: cycles binned by day and station 1`] = `
Map {
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"L/U #3\\",
}" => 6,
  "DayAndStation {
  \\"day\\": 2018-03-05T06:00:00.000Z,
  \\"station\\": \\"L/U #2\\",
}" => 3,
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"L/U #2\\",
}" => 6,
  "DayAndStation {
  \\"day\\": 2018-03-05T06:00:00.000Z,
  \\"station\\": \\"L/U #1\\",
}" => 5,
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"L/U #1\\",
}" => 13,
  "DayAndStation {
  \\"day\\": 2018-03-05T06:00:00.000Z,
  \\"station\\": \\"MC #1\\",
}" => 30,
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"MC #1\\",
}" => 35,
}
`;

exports[`bins actual cycles by day: occupied cycles binned by day and station 1`] = `
Map {
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"L/U #3\\",
}" => 8,
  "DayAndStation {
  \\"day\\": 2018-03-05T06:00:00.000Z,
  \\"station\\": \\"L/U #2\\",
}" => 3,
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"L/U #2\\",
}" => 6,
  "DayAndStation {
  \\"day\\": 2018-03-05T06:00:00.000Z,
  \\"station\\": \\"L/U #1\\",
}" => 5,
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"L/U #1\\",
}" => 13,
  "DayAndStation {
  \\"day\\": 2018-03-05T06:00:00.000Z,
  \\"station\\": \\"MC #1\\",
}" => 30,
  "DayAndStation {
  \\"day\\": 2018-03-04T06:00:00.000Z,
  \\"station\\": \\"MC #1\\",
}" => 35,
}
`;

exports[`creates points clipboard table: clipboard table 1`] = `
<div>
  <table>
    

    <thead>
      <tr>
        <th>
          Station
        </th>
        <th>
          Sun Mar 04 2018
        </th>
        <th>
          Mon Mar 05 2018
        </th>
      </tr>
    </thead>
    

    <tbody>
      

      <tr>
        <th>
          L/U #1
        </th>
        <td>
          13
        </td>
        <td>
          5
        </td>
      </tr>
      

      <tr>
        <th>
          L/U #2
        </th>
        <td>
          6
        </td>
        <td>
          3
        </td>
      </tr>
      

      <tr>
        <th>
          MC #1
        </th>
        <td>
          35
        </td>
        <td>
          30
        </td>
      </tr>
      

    </tbody>
    

  </table>
</div>
`;
